### DISCUSSION TITLE
Trying to figure out parent_dir

### DISCUSSION THREAD

**Discussion by daenney** [gokrazy/gokrazy#202] [Q&A] [answered]
I've been trying to figure out where I'm going wrong with gokrazy/tools#57 and how `parent_dir` is intended to be used. It sounds like the value of `parent_dir` may include some additional state but that so far doesn't seem to be the case, so I'm unclear what its purpose it. But without setting it, `gok new` will create the `$name/config.json` in `~/gokrazy`.

As I tried to outline in gokrazy/tools#57, I want to create a `config.json` plus some supporting documentation so that others can take that, tweak it a little and flash a Pi. I'd like to commit those to a new repo and push them out, so I'm trying to avoid it getting created under `~/gokrazy`.

I've tried the following, but the example won't build:

```
~/D/github.com ❯❯❯ mkdir parent-dir
~/D/github.com ❯❯❯ cd parent-dir/
~/D/g/parent-dir ❯❯❯ mkdir gotosocial
~/D/g/parent-dir ❯❯❯ cd gotosocial/
~/D/g/p/gotosocial ❯❯❯ gok --parent_dir .. new -i gotosocial
gokrazy instance configuration created in ../gotosocial/config.json
(Use 'gok -i gotosocial edit' to edit the configuration interactively.)

Use 'gok -i gotosocial add' to add packages to this instance

To deploy this gokrazy instance, see 'gok help overwrite'
~/D/g/p/gotosocial ❯❯❯ ls
breakglass.authorized_keys  config.json
```

At this point I only tried to build the given example, using:
<details><summary>build with --parent_dir ..</summary>

```
~/D/g/p/gotosocial ❯❯❯ gok --parent_dir .. overwrite -i gotosocial --root root.squashfs
gokrazy gok gcd4073 on GOARCH=amd64 GOOS=linux

Build target: CGO_ENABLED=0 GOARCH=arm64 GOOS=linux
Build timestamp: 2023-06-22T16:25:13+02:00
Loading system CA certificates from /etc/ssl/certs/ca-certificates.crt
Building 4 Go packages:

  github.com/gokrazy/fbstatus

  github.com/gokrazy/hello

  github.com/gokrazy/serial-busybox

  github.com/gokrazy/breakglass
    will be started with command-line flags
      from ../gotosocial/config.json
      last modified: 2023-06-22T16:24:02+02:00 (1m11s ago)

[building (go compiler)]go: finding module for package github.com/gokrazy/kernel
go: downloading github.com/gokrazy/kernel v0.0.0-20230622035841-e953d5bc6ad4
go: finding module for package github.com/gokrazy/firmware
go: finding module for package github.com/gokrazy/rpi-eeprom
go: finding module for package github.com/gokrazy/gokrazy/cmd/dhcp
go: finding module for package github.com/gokrazy/gokrazy/cmd/ntp
go: finding module for package github.com/gokrazy/gokrazy/cmd/randomd
go: finding module for package github.com/gokrazy/gokrazy/cmd/heartbeat
go: finding module for package github.com/gokrazy/fbstatus
go: finding module for package github.com/gokrazy/hello
go: finding module for package github.com/gokrazy/serial-busybox
go: finding module for package github.com/gokrazy/breakglass
go: finding module for package github.com/gokrazy/gokrazy
[done] in 22.04s                        

Including extra files for Go packages:

  github.com/gokrazy/breakglass
    will include extra files in the root file system
      from ../gotosocial/breakglass.authorized_keys
      last modified: 2023-06-22T16:24:02+02:00 (1m33s ago)

go: finding module for package github.com/gokrazy/gokrazy
Including loadable kernel modules from:
/home/daenney/.cache/go/pkg/mod/github.com/gokrazy/kernel@v0.0.0-20230622035841-e953d5bc6ad4/lib/modules
2023/06/22 16:25:36 Error: build directory "../gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/ntp" does not exist in "/home/daenney/Development/github.com/parent-dir/gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/dhcp"
Try 'gok -i gotosocial add github.com/gokrazy/gokrazy/cmd/ntp' followed by an update.
Afterwards, your 'gok sbom' command should work: chdir ../gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/ntp: no such file or directory
```
</details>

The build failure is what I'm puzzled about:
```
2023/06/22 16:25:36 Error: build directory "../gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/ntp" does not exist in "/home/daenney/Development/github.com/parent-dir/gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/dhcp"
Try 'gok -i gotosocial add github.com/gokrazy/gokrazy/cmd/ntp' followed by an update.
Afterwards, your 'gok sbom' command should work: chdir ../gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/ntp: no such file or directory
```

I'm not sure why this occurs, or why it seems to be looking for `cmd/ntp` inside `cmd/dhcp`?

If I go stand somewhere else in the filesystem entirely, and use the full path to `parent-dir` instead then the build works:
<details><summary>same command, but from /tmp with full path to parent-dir</summary>

```
~/D/g/parent-dir ❯❯❯ cd /tmp
/tmp ❯❯❯ gok --parent_dir /home/daenney/Development/github.com/parent-dir overwrite -i gotosocial --root root.squashfs
gokrazy gok gcd4073 on GOARCH=amd64 GOOS=linux

Build target: CGO_ENABLED=0 GOARCH=arm64 GOOS=linux
Build timestamp: 2023-06-22T16:32:14+02:00
Loading system CA certificates from /etc/ssl/certs/ca-certificates.crt
Building 4 Go packages:

  github.com/gokrazy/fbstatus

  github.com/gokrazy/hello

  github.com/gokrazy/serial-busybox

  github.com/gokrazy/breakglass
    will be started with command-line flags
      from /home/daenney/Development/github.com/parent-dir/gotosocial/config.json
      last modified: 2023-06-22T16:24:02+02:00 (8m12s ago)

[done] in 1.42s                        

Including extra files for Go packages:

  github.com/gokrazy/breakglass
    will include extra files in the root file system
      from ../gotosocial/breakglass.authorized_keys
      last modified: 2023-06-22T16:24:02+02:00 (8m14s ago)

Including loadable kernel modules from:
/home/daenney/.cache/go/pkg/mod/github.com/gokrazy/kernel@v0.0.0-20230622035841-e953d5bc6ad4/lib/modules

Feature summary:
  use GPT: true
  use PARTUUID: true
  use GPT PARTUUID: true

Creating root file system
[done] in 0.54s                           

Build complete!

To interact with the device, gokrazy provides a web interface reachable at:

	http://gokrazy:wBJIQcl1RhND0IFN2NxU@gotosocial/

In addition, the following Linux consoles are set up:

	1. foreground Linux framebuffer console on HDMI
```
</details>

If I try the same setup without using `--parent_dir` at all, so it ends up defaulting to `~/gokrazy` then things also work.
<details><summary>same commands as before, just without parent_dir</summary>

```
~ ❯❯❯ gok new -i gotosocial
gokrazy instance configuration created in /home/daenney/gokrazy/gotosocial/config.json
(Use 'gok -i gotosocial edit' to edit the configuration interactively.)

Use 'gok -i gotosocial add' to add packages to this instance

To deploy this gokrazy instance, see 'gok help overwrite'
~ ❯❯❯ gok -i gotosocial overwrite --root root.squashfs
gokrazy gok gcd4073 on GOARCH=amd64 GOOS=linux

Build target: CGO_ENABLED=0 GOARCH=arm64 GOOS=linux
Build timestamp: 2023-06-22T16:30:20+02:00
Loading system CA certificates from /etc/ssl/certs/ca-certificates.crt
Building 4 Go packages:

  github.com/gokrazy/fbstatus

  github.com/gokrazy/hello

  github.com/gokrazy/serial-busybox

  github.com/gokrazy/breakglass
    will be started with command-line flags
      from /home/daenney/gokrazy/gotosocial/config.json
      last modified: 2023-06-22T16:30:03+02:00 (17s ago)

[building (go compiler)]go: finding module for package github.com/gokrazy/kernel
go: finding module for package github.com/gokrazy/firmware
go: finding module for package github.com/gokrazy/rpi-eeprom
go: finding module for package github.com/gokrazy/gokrazy/cmd/dhcp
go: finding module for package github.com/gokrazy/gokrazy/cmd/ntp
go: finding module for package github.com/gokrazy/gokrazy/cmd/randomd
go: finding module for package github.com/gokrazy/gokrazy/cmd/heartbeat
go: finding module for package github.com/gokrazy/fbstatus
go: finding module for package github.com/gokrazy/hello
go: finding module for package github.com/gokrazy/serial-busybox
go: finding module for package github.com/gokrazy/breakglass
go: finding module for package github.com/gokrazy/gokrazy
[done] in 3.80s                        

Including extra files for Go packages:

  github.com/gokrazy/breakglass
    will include extra files in the root file system
      from /home/daenney/gokrazy/gotosocial/breakglass.authorized_keys
      last modified: 2023-06-22T16:30:03+02:00 (21s ago)

go: finding module for package github.com/gokrazy/gokrazy
Including loadable kernel modules from:
/home/daenney/.cache/go/pkg/mod/github.com/gokrazy/kernel@v0.0.0-20230622035841-e953d5bc6ad4/lib/modules

Feature summary:
  use GPT: true
  use PARTUUID: true
  use GPT PARTUUID: true

Creating root file system
[done] in 0.54s                           

Build complete!

To interact with the device, gokrazy provides a web interface reachable at:

	http://gokrazy:8Hr2Bj4x2eooopJHhpgs@gotosocial/

In addition, the following Linux consoles are set up:

	1. foreground Linux framebuffer console on HDMI
```
</details>

However, if I go into `~/gokrazy/gotosocial` now and try to build with `--parent_dir ..`, then I get the ntp package build error again

<details><summary>build from /home/daenney/gokrazy/gotosocial</summary>

```
~ ❯❯❯ cd ~/gokrazy/gotosocial/
~/g/gotosocial ❯❯❯ ls
breakglass.authorized_keys  builddir/  config.json
~/g/gotosocial ❯❯❯ gok --parent_dir .. -i gotosocial overwrite --root root.squashfs
gokrazy gok gcd4073 on GOARCH=amd64 GOOS=linux

Build target: CGO_ENABLED=0 GOARCH=arm64 GOOS=linux
Build timestamp: 2023-06-22T16:39:06+02:00
Loading system CA certificates from /etc/ssl/certs/ca-certificates.crt
Building 4 Go packages:

  github.com/gokrazy/fbstatus

  github.com/gokrazy/hello

  github.com/gokrazy/serial-busybox

  github.com/gokrazy/breakglass
    will be started with command-line flags
      from ../gotosocial/config.json
      last modified: 2023-06-22T16:30:03+02:00 (9m3s ago)

[done] in 1.42s                        

Including extra files for Go packages:

  github.com/gokrazy/breakglass
    will include extra files in the root file system
      from /home/daenney/gokrazy/gotosocial/breakglass.authorized_keys
      last modified: 2023-06-22T16:30:03+02:00 (9m4s ago)

Including loadable kernel modules from:
/home/daenney/.cache/go/pkg/mod/github.com/gokrazy/kernel@v0.0.0-20230622035841-e953d5bc6ad4/lib/modules
2023/06/22 16:39:08 Error: build directory "../gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/ntp" does not exist in "/home/daenney/gokrazy/gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/dhcp"
Try 'gok -i gotosocial add github.com/gokrazy/gokrazy/cmd/ntp' followed by an update.
Afterwards, your 'gok sbom' command should work: chdir ../gotosocial/builddir/github.com/gokrazy/gokrazy/cmd/ntp: no such file or directory
```
</details>

Basically, I'm pretty convinced I'm holding it wrong, but I'm also unsure how to hold it right and I can't figure it out from the docs :sweat_smile:.

I'll work from the instance in `~/gokrazy` in the mean time, but I'd like to understand how `parent_dir` is intended to function so I can get things to appear where I expect them to and have the `overwrite` command complete successfully.

**Comment by stapelberg** [comment, marked as answer]
> I've been trying to figure out where I'm going wrong with [gokrazy/tools#57](https://github.com/gokrazy/tools/issues/57) and how `parent_dir` is intended to be used. It sounds like the value of `parent_dir` may include some additional state but that so far doesn't seem to be the case, so I'm unclear what its purpose it.

There’s no additional state in the parent directory. It is only used whenever an instance name (`-i gotosocial`) needs to be translated into an absolute directory path. 

I’m thinking there might just be a bug (or two), but I don’t have time to look at this right now. Feel free to poke around and see if you can figure it out in the meantime.

It might be best to keep the entire discussion in https://github.com/gokrazy/tools/issues/57 instead of two places.

**Reply by daenney** [reply]
That's fair. I put it here because I got the distinct impression it was me doing something weird and didn't feel like that belonged in a bug report. Thanks for taking a look! I'll see if I can figure out where the path resolution goes wrong.
